### PR TITLE
Add Milestone to PR

### LINKED ISSUE #45: Add Milestone to PR
### Is your enhancement related to a problem? Please describe.

 Add Milestone to PR.
 
 Reference:
 - https://github.com/WordPress/gutenberg/tree/trunk/packages/project-management-automation/lib/tasks/add-milestone from https://github.com/WordPress/gutenberg/tree/trunk/packages/project-management-automation

### Designs

_No response_

### Describe alternatives you've considered

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by iamdharmesh**
### Description of the Change
Add auto-assigning milestone feature. 

This feature helps to automatically add milestone to Pull Requests. If the PR is connected to an issue with a milestone, the same milestone will be added to the PR. Otherwise, the next milestone from the available milestones will be assigned, sorted using version comparison.

Closes #45

### How to test the Change
1. Create a [workflow file](https://github.com/iamdharmesh/action-pr-validator/pull/46/files) using this action and pointing to this branch. (`uses: 10up/action-pr-automator@enhancement/45`)
2. Create one sample issue, assign a milestone to it, and raise a sample PR that fixes the created issue.
3. Verify the same milestone added to PR.
4. Create one more sample issue without milestone and raise a sample PR that fixes the created issue.
5. Verify that milestone added to PR from the available milestones, sorted using version comparison. (For example, if `1.2.1`, `1.3.0` and `Future Release` milestones are available then `1.2.1` will be added.)

### Credits
Props @iamdharmesh, @jeffpaul.

### Checklist:

- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [ ] All new and existing tests pass.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -22,6 +23,7 @@ This GitHub Action Helps with the following operations:
 | Key | Default | Description |
 | --- | ------- | ----------- |
 | assign-pr | true | Wether assign PR to reporter |
+| add-milestone | true | Wether add milestone to PR |
```
Comment (faisal-alvi) [inline README.md]:
```suggestion
| add-milestone | true | Whether to add a milestone to the PR |
```

File: README.md [hunk 2]
```diff
@@ -22,6 +23,7 @@ This GitHub Action Helps with the following operations:
 | Key | Default | Description |
 | --- | ------- | ----------- |
 | assign-pr | true | Wether assign PR to reporter |
```
Comment (faisal-alvi) [inline README.md]:
```suggestion
| assign-pr | true | Whether assign PR to reporter |
```

File: action.yml [hunk 3]
```diff
@@ -8,6 +8,10 @@ inputs:
     description: "Wether assign PR to reporter"
     required: false
     default: true
+  add-milestone:
+    description: "Wether add milestone to PR"
```
Comment (faisal-alvi) [inline action.yml]:
```suggestion
    description: "Whether add milestone to PR"
```

File: action.yml [hunk 4]
```diff
@@ -8,6 +8,10 @@ inputs:
     description: "Wether assign PR to reporter"
```
Comment (faisal-alvi) [inline action.yml]:
```suggestion
    description: "Whether assign PR to reporter"
```

**Review by faisal-alvi — COMMENTED** [review]
@iamdharmesh thanks for the great work here.

1. I've followed the instructions from the PR description to test and tried to test this PR at https://github.com/faisal-alvi/action-pr-automator/pull/3 but it does not work for me. Can you please check and let me know if I missed anything?

image
[link](https://github.com/faisal-alvi/action-pr-automator/actions/runs/4162713916/jobs/7202224970)

2. Also, even though it failed, the job was successful, I assume the job should be considered failed.

**Comment by iamdharmesh** [comment]
Hi @faisal-alvi, 

Thanks for checking on this and providing feedback.

> 1. I've followed the instructions from the PR description to test and tried to test this PR at https://github.com/faisal-alvi/action-pr-automator/pull/3 but it does not work for me. Can you please check and let me know if I missed anything?

The connected issue has milestone `1.0.0` and the current version of your repo is `1.0.1`. So, it ignores adding milestone to the PR. Could you please try after adding milestone something like `1.0.2` or `1.1.0` to the issue? (version comparison is added to select the best possible milestone in case of multiple connected issues with different milestones)

> 2. Also, even though it failed, the job was successful, I assume the job should be considered failed.

The job fails only if the PR description does not pass the validation, this feature is to automate stuff. So, if the action does fail to perform action, it can be done manually. Making this job fail may prevent us from merging PR.


Thanks

**Review by faisal-alvi — APPROVED** [review]
@iamdharmesh thanks for the efforts! I have tested https://github.com/iamdharmesh/action-pr-validator/pull/51 and the milestone added by the bot successfully! 🚀

**Pull request merged**